Let RpcAgent::send() return JitFuture

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50236 Apply clang-format to rpc cpp files
* #50029 Completely remove FutureMessage type
* #50028 Completely Remove FutureMessage from RPC agents
* #50027 Completely Remove FutureMessage from RPC cpp tests
* #50026 Remove FutureMessage from RPC request callback logic
* #50025 Completely remove FutureMessage from FaultyProcessGroupAgent
* #50024 Remove FutureMessage from sender TensorPipeAgent
* #50023 Remove FutureMessage from sender ProcessGroupAgent
* #50020 Completely remove FutureMessage from distributed autograd
* #50005 Completely remove FutureMessage from RPC TorchScript implementations
* #50004 Completely remove FutureMessage from RRef Implementations
* #49995 Replace FutureMessage with ivalue::Future in RpcAgent retry logic
* #49960 Replace FutureMessage with ivalue::Future in RRefContext
* #49927 Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*
* **#49906 Let RpcAgent::send() return JitFuture**

This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

mrshenli committed:
Let RpcAgent::send() return JitFuture

This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

[ghstack-poisoned]

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

[ghstack-poisoned]

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -259,6 +259,38 @@ class TORCH_API RpcAgent {
+    auto jitFuture = std::make_shared<JitFuture>(at::AnyClassType::get());
+
+    fm->addCallback(
+        [jitFuture, fm]() mutable {
```
I remember last time I worked on C++ futures API, Jeremy had a comment here: https://github.com/pytorch/pytorch/pull/40326#discussion_r449289403 about avoiding (possible) refcycles and suggesting to use a weak_ptr. 

The motivation is that sometimes, the callbacks are never run (for various reasons). In this case, we have a copy of the shared_ptr so an additional reference to it, so the pointed to memory is never freed due to the ref cycle.

Ultimately in that PR we went with using weakref instead of the shared_ptr copy, so just wanted to call it out in case we want to do that here as well.

pritamdamania87 reviewed (commented):
Left a few comments on the internal diff.

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -841,6 +841,11 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+  // NB: need to call torch::class_ to register Message in the map returned by
+  // c10::getCustomClassTypeMap(). Otherwise, Message cannot be wrapped within
+  // an IValue.
+  static const auto message = torch::class_<Message>("rpc", "_Message");
```
Hey @wanchaol, are there plans to extract the custom type registration logic from `torch::class_` to convert it into a utility function, so that c++ only IValue custom types no longer need to be bound here?

wanchaol replied:
Sorry for the late reply. is this still a valid ask? I think if this is only for C++ used, we don't need to register/bound the class into python or TorchScript, just make the C++ class inherit `torch::CustomClassHolder` is enough

mrshenli replied:
Hey @wanchaol, thanks for the comments. Yep, this is still relevant. I found that if I don't call `torch::class_`, `IValue(c10::make_intrusive<Message>(std::move(msg)))` would crash. I traced the code and found that it hits this line:

https://github.com/pytorch/pytorch/blob/71ca600af93b4ad58ca6fc083fa5d42a56b43194/aten/src/ATen/core/ivalue.h#L963

I did a quick search, and looks like `torch::class_` is only place that inserts new types:

https://github.com/pytorch/pytorch/blob/00a3add425d305f79a413449d570e8f08e7a4c81/torch/custom_class.h#L88-L91

Please let me know if there are better suggestions. Thanks!

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]

mrshenli reviewed (commented):
This PR has internal changes

lw commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -157,7 +157,7 @@ class TORCH_API RpcAgent {
   // completed when the response arrives. For other message types, the Future
   // should be ignored by the caller.
-  virtual std::shared_ptr<FutureMessage> send(
+  virtual std::shared_ptr<JitFuture> send(
```
I've mostly seen JitFuture being passed around through an intrusive_ptr, rather than a shared_ptr. I guess that's somewhat more efficient (not sure how much). Should we consider it here too?

mrshenli replied:
@wanchaol @gmagogsfm is there any perf penalty if we use `ivalue::Future` with `std::shared_ptr`?

wanchaol replied:
There shouldn't be too much perf penalty, it's just that shared_ptr store the refcount in a separate memory space instead of embedded in the object, so from cache locality prospective intrusive_ptr should be a bit faster. We passed around JitFuture through intrusive_ptr also for the reason of making it to be able to held by `IValue` so that it's easier to bind to TorchScript if needed.

mrshenli replied:
I see, thanks for the explanation!

lw commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -104,6 +104,18 @@ Message createExceptionResponse(const std::string& exceptionStr, int64_t id) {
+// c10::getCustomClassTypeMap(). Otherwise, Message cannot be wrapped within
+// an IValue.
+// NB: add this line here instead of in rpc/init.cpp because 1) we have C++ only
+// tests that won't run rpc/init.cpp; 2) Message is not mean to be visible from
```
```suggestion
// tests that won't run rpc/init.cpp; 2) Message is not meant to be visible from
```

lw reviewed (approved):
The OSS changes look good to me. Although I'm not qualified on the JIT custom classes, so let's get a review from someone else too.

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RpcAgent::send() return JitFuture"


This commit modifies RPC Message to inherit from `torch::CustomClassHolder`,
and wraps a Message in an IValue in `RpcAgent::send()`.

Differential Revision: [D25719518](https://our.internmc.facebook.com/intern/diff/D25719518)

[ghstack-poisoned]
